### ISSUE TITLE
win_updates stuck

### ISSUE THREAD

**Issue by akhmarov** [ansible-collections/ansible.windows#253]
##### SUMMARY
win_updates stuck

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
win_updates

##### ANSIBLE VERSION

```paste below
ansible [core 2.11.3] 
  config file = None
  configured module search path = ['/Users/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/4.3.0/libexec/lib/python3.9/site-packages/ansible
  ansible collection location = /Users/user/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.9.6 (default, Jun 29 2021, 05:25:02) [Clang 12.0.5 (clang-1205.0.22.9)]
  jinja version = 3.0.1
  libyaml = True
```

##### COLLECTION VERSION

```paste below
# /Users/user/.ansible/collections/ansible_collections
Collection      Version
--------------- -------
ansible.windows 1.7.0  

# /usr/local/Cellar/ansible/4.3.0/libexec/lib/python3.9/site-packages/ansible_collections
Collection      Version
--------------- -------
ansible.windows 1.7.0  
```

##### CONFIGURATION

```paste below
-
```

##### OS / ENVIRONMENT

Client Version: macOS 11.4 (20F71)
Target OS version: Windows Server 2019 patched to latest approved updates via corporate WSUS

##### STEPS TO REPRODUCE

```yaml
- name: 'Windows :: Installing critical and security updates'
  win_updates:
    category_names:
      - CriticalUpdates
      - SecurityUpdates
    state: installed
    reboot: yes
```

no_proxy="*" ansible-playbook -vvvv -i hosts.ini deploy.yml --user user@domain.tld --ask-pass

##### EXPECTED RESULTS

win_updates completes successfully

##### ACTUAL RESULTS

win_updates stuck in infinite loop; playbook stops for day or longer (cannot check further)

```paste below
TASK [Windows :: Installing critical and security updates] ***********************************************************************************************************************************************************************************
task path: /Users/user/Documents/Git/ansible_play/windows/windows_update.yml:18
redirecting (type: action) ansible.builtin.win_updates to ansible.windows.win_updates
redirecting (type: action) ansible.builtin.win_updates to ansible.windows.win_updates
<server.domain.tld> ESTABLISH WINRM CONNECTION FOR USER: user@domain.tld on PORT 5986 TO server.domain.tld
EXEC (via pipeline wrapper)
<server.domain.tld> PUT "/Users/user/.ansible/tmp/ansible-local-50741yup8vw7l/tmpbe5bu9nm/poll.ps1" TO "C:\Users\user\AppData\Local\Temp\ansible-tmp-1626892064.055095-50746-22575345614491\poll.ps1"
<server.domain.tld> PUT "/Users/user/.ansible/tmp/ansible-local-50741yup8vw7l/tmp9ke9guk3/cancel.ps1" TO "C:\Users\user\AppData\Local\Temp\ansible-tmp-1626892064.055095-50746-22575345614491\cancel.ps1"
<server.domain.tld> Running win_updates - round 1
<server.domain.tld> Starting update task
Using module file /Users/user/.ansible/collections/ansible_collections/ansible/windows/plugins/modules/win_updates.ps1
Pipelining is enabled.
EXEC (via pipeline wrapper)
```

If we look at target server's task scheduler we found that task "ansible-ansible.windows.win_updates". History will contain:
7/21/2021 9:27:48 PM Task Scheduler launched "{8c1ef1cd-09e9-4b5e-ab3a-7f9b660027bb}"  instance of task "\ansible-ansible.windows.win_updates"  for user "user" .
7/21/2021 9:27:48 PM Task Scheduler failed to start "\ansible-ansible.windows.win_updates" task for user "user". Additional Data: Error Value: 2147943785.

According to task scheduler command to run is "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" with argument '-ExecutionPolicy ByPass -NoProfile -NonInteractive -File "C:\Users\user\AppData\Local\Temp\ansible-tmp-1626892064.055095-50746-22575345614491\.\task.ps1" -Id "ansible-ansible.windows.win_updates-4bf46b7e-37c2-4a7a-a8eb-0ef59ccaab8b" -LogPath "C:\Users\user\AppData\Local\Temp\ansible-tmp-1626892064.055095-50746-22575345614491\."'

If we try to run this command manually C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -ExecutionPolicy ByPass -NoProfile -NonInteractive -File "C:\Users\user\AppData\Local\Temp\ansible-tmp-1626892064.055095-50746-22575345614491\.\task.ps1" -Id "ansible-ansible.windows.win_updates-4bf46b7e-37c2-4a7a-a8eb-0ef59ccaab8b" -LogPath "C:\Users\user\AppData\Local\Temp\ansible-tmp-1626892064.055095-50746-22575345614491\." it will not return any output after 1 hour

None new log files found under C:\Users\user\AppData\Local\Temp\ansible-tmp-1626892064.055095-50746-22575345614491\.

task.ps1 file attached
[task.ps1.zip](https://github.com/ansible-collections/ansible.windows/files/6858117/task.ps1.zip)

**Comment by jborean93** [comment]
Thanks for the detailed information, `task.ps1` is just a generic script that is used to start the windows update code. It uses a named pipe to get the code to run as a scheduled task but if you don't send it anything it's going to wait indefinitely. This is why running it manually will just hang because you also aren't sending the data over the pipe specified for it to continue.

Can you add `log_path` to the `win_updates` options and see if the module actually starts adding information to the log. This is an easy way to determine what the module is actually doing during execution and check if it's hanging or just taking a really long time. Are there any other files in the directory you found `task.ps1` in?

The error you you shared relates to the HRESULT 0x0x80070569 which means

> Logon failure: the user has not been granted the requested logon type at this computer

If the connecting user `user@domain.tld` an administrator on that host? Does it have the rights to log on as a batch user. My guess is either one of those 2 are false and that causes the task to never start and the initial run to perceptually wait for a client to connect. Running the [win_whoami](https://docs.ansible.com/ansible/latest/collections/ansible/windows/win_whoami_module.html) task will show you whether the user is an admin when logging in with Ansible and whether it has the `SeBatchLogonRight` user right required to run as a scheduled task.

**Comment by akhmarov** [comment]
Hello Jordan

Thanks for your answer. Indeed adding SeBatchLogonRight permissions to ansible_user before installing Windows Updates and removing it just after fixed my problem. Would like to add that earlier this year there was no need to do such actions at all. I used the same DOMAIN account without SeBatchLogonRight permissions and everything was fine

**Comment by jborean93** [comment]
> Thanks for your answer. Indeed adding SeBatchLogonRight permissions to ansible_user before installing Windows Updates and removing it just after fixed my problem

Thanks for confirming that was the problem.

> Would like to add that earlier this year there was no need to do such actions at all

Yes this is my fault. I've recently refactored the `win_updates` plugin to change how it operates to try and be more resiliant towards network failures. One of the changes is that the actual update code is run under a scheduled task whereas before that was an opt-in behavior. Having the module just hang there with no indication what is happening is definitely not ideal so I've made a change https://github.com/ansible-collections/ansible.windows/pull/255 that makes it a bit smarter when waiting for the code to start. It also adds a fallback to run the task as `SYSTEM` which negates the need for this workaround in your case.
